### DISCUSSION TITLE
Severe IDE slowdowns due to console contents

### DISCUSSION THREAD

**Discussion by jonahgraham** [eclipse/lsp4e#419] [Q&A]
We had a report on Eclipse Platform (see https://github.com/eclipse-platform/eclipse.platform.swt/issues/539) that the IDE was hanging at startup. The cause appeared to be something about the content of the console, probably the LSP4E one.

In diagnosing the problem the OP reported that the console had 90k lines of output. As a possible workaround OP has now disabled LSP4E console and limited console output and reports no more hangs for the last day or two.

Does anyone know if this is expected behaviour, do we already have knowledge about the performance in this area?

Has anyone else seen hanging related to the LSP4E console, in particular on Windows with captured stack traces that look like this?

```java
"main" #1 prio=6 os_prio=0 cpu=16956234.38ms elapsed=30861.56s tid=0x00000148ee5f1a20 nid=30224 runnable  [0x0000003a16b1d000]
   java.lang.Thread.State: RUNNABLE
        at org.eclipse.swt.internal.win32.OS.ScriptPlace(Native Method)
        at org.eclipse.swt.graphics.TextLayout.shape(TextLayout.java:3760)
        at org.eclipse.swt.graphics.TextLayout.computeRuns(TextLayout.java:405)
        at org.eclipse.swt.graphics.TextLayout.getBounds(TextLayout.java:1657)
        at org.eclipse.swt.custom.StyledTextRenderer.calculate(StyledTextRenderer.java:299)
```
(full stack available in https://github.com/eclipse-platform/eclipse.platform.swt/issues/539)

**Comment by mickaelistria** [comment]
Do you have an idea what the content of the console would be? Is it LSP protocol-level messages being logged by LSP4E as initially intended for debug or is it the Language Server being actually verbose and sending tons of `logMessage`?
Do you have any idea what language server is involved?

It it's protocol-level messages, I personally have stopped using them long ago because they got to big to be processed in the console. So if it's the cause and no other committer uses it; we may just remove the option.

**Reply by jonahgraham** [reply]
@davidmichaelkarr are you able to answer @mickaelistria's questions?

**Comment by davidmichaelkarr** [comment]
Later this morning, I will re-enable the lsp4e console and see what I see.  Hopefully it won't cause Eclipse to hang again.  The one thing I DO remember from viewing the lsp4e log entries is an occasional "Invalid location" exception.

**Comment by davidmichaelkarr** [comment]
I enabled the lsp4e logging in 2022-12, and I'm attaching the content for about an hour's worth.  I've gzipped it because the file is pretty big.
[lsp4e.tar.gz](https://github.com/eclipse/lsp4e/files/10538688/lsp4e.tar.gz)

If it matters, I didn't see any "Invalid location" exceptions in this log.

**Reply by jonahgraham** [reply]
Thanks @davidmichaelkarr.

@mickaelistria I think the log answers your questions:

**Reply by jonahgraham** [reply]
> Do you have an idea what the content of the console would be?

Protocol messages

> Do you have any idea what language server is involved?

org.eclipse.languageserver.languages.springboot

> So if it's the cause and no other committer uses it; we may just remove the option.

Sounds reasonable, especially with what I note below:

---

Interestingly the log contains a single line that is nearly 2MB and does not render properly with mousepad or Eclipse on Linux:

Mousepad:

![image](https://user-images.githubusercontent.com/679236/215568001-fe2b2752-5f88-4b96-abd9-14fec105d17c.png)

Eclipse:

![image](https://user-images.githubusercontent.com/679236/215568699-f82a72cb-1c42-4288-b997-bf1713ddb42c.png)

In vscode it truncates the line, and pressing Show More hangs VS Code.

![image](https://user-images.githubusercontent.com/679236/215568475-3d4af344-4963-46b0-8ca9-48443e0f4316.png)

I'll try on Windows at some point to see what the line does there, especially with wordwrapping.

**Reply by mickaelistria** [reply]
Thanks for the details, I've opened https://github.com/eclipse/lsp4e/issues/424

**Reply by martinlippert** [reply]
The overly large message from the spring boot language server is indeed not optimal (even beyond the trouble it might cause here when logging it to the console), so we have an issue open to address that from the spring boot language server side as well.

**Comment by davidmichaelkarr** [comment]
I did not notice any obvious issues with rendering the text in that log, at least in Emacs.  I found it curious that there are often a few blank lines between log entries, and I do see that honkin' big one, but Emacs has no trouble rendering it.

I also find it curious that I now have set a limit on the length of the console buffer at 200k, but that doesn't appear to apply to this. Clearly, this log is more than 200k.

I also note that I haven't seen the total hang for a while now.

**Reply by jonahgraham** [reply]
vi also has no trouble rendering it, vi and emacs AFAIK both don't use the same script rendering engine as Eclipse and many other UI programs.

> I also note that I haven't seen the total hang for a while now.

Do you still have wordwrap off?

**Comment by davidmichaelkarr** [comment]
It is presently off.  I tried turning it on, and was briefly freaked when it took a few seconds for the hourglass to go away.  I turned it back off.

I'm just guessing, but it feels like the hang isn't due to a single factor, but a combination of factors.

So does the "Console buffer size" field in the "Console" page of preferences not apply to the lsp4e console?  I don't think I've seen the hang since I set the limit of 200k, but my lsp4e console is much more than 200k.  However, I wonder about the Maven console.  If I didn't have a limit set there, the content in the Maven console would continue to grow, and I have quite a few Maven projects open.  This is obviously offtopic for this repo, but I'm just trying to get the overall problem solved.

**Comment by rubenporras** [comment]
We use it for testing purposes, we don plan to use it on production. Removing the log to console feature would be fine for, as one can open the file in an external editor, but I actually find it handy, that one can have it on the console, always for debugging purposes. We could maybe remove the UI dialog to activate it, but keep the option so that it can be set with plugin properties.

**Reply by ahmedneilhussain** [reply]
I also find logging to console invaluable for dev work and debugging, so I would be sorry to lose it as a dev option, even if not exposed to users. Logging to a file would be ok so long as it's easy to configure.

**Reply by mickaelistria** [reply]
OK, so let's keep things as it.
@davidmichaelkarr I don't know how you got the messages logged in console, but this is not something you should do. Just disable logging to console in Prefrences > Langauge Servers > Logs.

**Comment by davidmichaelkarr** [comment]
Ok, well, I had enabled lsp4e logging because I was concerned about the perf problems I was seeing with lsp4e, to see if I could see any information that might be useful to you. Just now, I disabled lsp4e logging, restarted, and now my laptop's cpu has been running flat out for several minutes, with the LSP process the top of the list, and Eclipse doing nothing. So, I killed the LSP process, and then it started back up again, using max cpu again.

**Reply by martinlippert** [reply]
@davidmichaelkarr If this is the Spring Boot Language Server process consuming all that CPU, I would suggest to report that at here (https://github.com/spring-projects/sts4/issues/) as well. I would be very interested in investigating what exactly is causing all that CPU consumption on the language server side.

**Reply by davidmichaelkarr** [reply]
Ok. I think it is. I had just turned off "Start Language Server at startup if Spring Boot is a dependency" and restarted, and curiously I still saw "BootLanguageServerBootApp" running when I looked at processes, but it only took a modest amount of cpu.  I'll try turning that on again and see what I see, and if it's bad, I'll report it in sts4.

**Reply by davidmichaelkarr** [reply]
At least this test didn't show extraordinary cpu usage.  It was high for a few seconds, but went down shortly after that. As usual, bugs tend to hide when you go looking for them.

**Reply by ahmedneilhussain** [reply]
Particularly performance bugs, I find. Everything grinds to a halt so badly that you can barely fire up `top` or `TaskManager` to see what on Earth is going on, but when the diagnostic tool finally starts it is when the machine is idle again.

It's like when the teacher leaves the classroom temporarily, prompting bedlam; yet when the headmaster sticks his head round the door all the pupils are sitting at their desks wearing expressions of angelic innocence.

**Comment by davidmichaelkarr** [comment]
I then reenabled lsp4e logging and restarted, and now the lsp process is barely using cpu.  Talk about side effects.
